joomla-projects/gsoc21_core-enhancements
# Core Enhancement 3 - Merge Featured into Articles

The purpose of this PR is to track all the file changes and commits made. Mentors can review the changes and provide feedback.
I will update the description as and when I push more commits.

Plan Document: https://docs.google.com/document/d/1Pl8JGa2hkYkmJzQOn9_mS8a4imDmqc2a/edit#heading=h.t5tri5tytv7q

<hr>

- [x] Drop-down selector in Articles view
- [x] Merge Model
- [x] Merge tmpl files
- [x] Merge View
- [x] Merge Controller
- [x] Update all links pointing to `view=featured`
- [ ] Test Sample Data
- [ ] Remove Featured files
- [x] Update Link of Featured Articles Menu Item

Edit: Merge Table has been removed from the above list because I feel that both the files (in [/src/Table/](https://github.com/joomla-projects/gsoc21_core-enhancements/tree/4.1-dev/administrator/components/com_content/src/Table)) play an essential role and they must be different 

## Summary of Changes

### [A] New Selector Dropdown
A searchtools selector is added that can be used to toggle between:

Option | Value
------------ | -------------
All  | ""
Unfeatured  | 0
Featured  | 1

![dropdown](https://user-images.githubusercontent.com/53610833/128350490-c5d83ab1-2ffb-42f7-8ecc-27cde5d658c5.gif)

<br>

### [B] Model Merged
- A new method to return the featured selector filter parameter using the getUserState from request
```php
$featured = $this->getUserStateFromRequest($this->context . '.featured', 'featured', '');
```
- The advantage of the above is that we can also pass the value as a GET param to the URL which helps us to make redirect URLs that can initialize the value of this dropdown
- `$featured` variable is used as a condition to manipulate the query to adjust to Featured

<br>

### [C] Templates Merged
- We use `state->get('featured')` to conditonally render the template code as per featured or not.

<br>

### [D] Views Merged
- We use `state->get('featured')` to conditonally render the toolbar options

<br>

### [E] Controller Merged
- Merged the delete function from FeaturedController

<br>

## Testing Instructions

- Visit the Articles View and ensure that everything works as it did before.
- Test workflows and filters too
- Ensure that the tmpl and view (and all other MVC files) are common for both featured and articles

<br>

## Expected result AFTER applying this Pull Request

Everything works normally
